include/boost/geometry/io/wkt/write.hpp

```diff
@@ -113,12 +160,13 @@ struct double_closing_parenthesis
 template <typename Point, typename Policy>
 struct wkt_point
 {
-    template <typename Char, typename Traits>
-    static inline void apply(std::basic_ostream<Char, Traits>& os, Point const& p, bool)
+    template <typename Output>
+    static inline void apply(Output& out, Point const& p, int significant_digits, bool)
```

Comment:
No, please don't do this.

There is more to stream than significant digits alone.

The original implementation should stay untouched, we should just be able to stream using any stream operator there is.

I thought that a wrapper should be build on top, such that we could stream to a string. But that is not this implementation.

Reply:
> No, please don't do this.

Do you mean having the `int significant_digits`  which is ignored in all cases but `std::string`?

Reply:
@mloskot that was my first proposal. But after that we agreed that it would be better create the string from the base and not just casting it from the stringstream to a string as mentioned [here](https://github.com/boostorg/geometry/issues/654#issuecomment-592458473). But in my opinion everybody should agree on one after hearing all arguments. So can you explicitly state what is the better approach according to you @barendgehrels ?

Reply:
@barendgehrels the stream-based implementation should not have been touched, hence my objection above about additional parameters to `bg::wkt()`.

@barendgehrels could you clarify what is the exact problem with this PR?
Is the problem that passing a dummy `significant_digits` into stream-based implementation is not elegant or that passing `significant_digits` into `bg::to_wkt()` at all is not elegant or something else?
Is the argument that there should be more parameters in `bg::to_wkt()` or no parameters at all or that there should be no `bg::to_wkt()`?

Note that if we had this simple version of `bg::to_wkt()` function mentioned by @mloskot then we'd have the same problems. It's only a matter of implementation.

@barendgehrels how about removing `significant_digits` parameter and handle the coordinates always in a default way in `bg::to_wkt()`?

Reply:
> 
> 
> @barendgehrels
> 
> > No, please don't do this.
> 
> Do you mean having the `int significant_digits` which is ignored in all cases but `std::string`?
> 
> > I thought that a wrapper should be build on top, such that we could stream to a string.
> 
> Do you mean expected implementation is something like this?
> 
> ```c++
> template <typename Geometry>
> inline std::string to_wkt(Geometry const& geometry, int significant_digits)
> {
>     std::stringstream ss;
>     ss.precision(significant_digits);
>     ss << geometry;
>     return ss.str();
> }
> ```

Yes, I had expected exactly this.

Reply:
Thanks for your reactions. The significant_digits not the only issue, the problem is that I don't see a reason to change the implementation at all. If there are good reasons - OK of course. But if significant_digits is the reason, and streaming to a string - that is sooo easy to solve in a different way (see proposal of Mateusz) that I really don't see the reason to add this extra complexity.

Reply:
> 
> 
> @mloskot that was my first proposal. But after that we agreed that it would be better create the string from the base and not just casting it from the stringstream to a string as mentioned [here](https://github.com/boostorg/geometry/issues/654#issuecomment-592458473). But in my opinion everybody should agree on one after hearing all arguments. So can you explicitly state what is the better approach according to you @barendgehrels ?

First of all, thanks of course for your efforts! And sorry if my comment came in late an unexpectedly.

To be honest, I missed this suggestion earlier, and I wonder why it would be better. Is the performance better? Both approaches use stringstream internally, so I don't expect significant differences. And adding significant_digits only quite limits it.

OK - so you followed the suggestion and now I call you back... I like this comment of Mateusz there ("My rule of thumb is to always start with option that is the least intrusive to the existing code.") and I wonder why we didn't follow it...

Reply:
@barendgehrels this was my suggestion based on the fact that I've made a simple benchmark and the result was that `std::string::append()` and `std::string::operator+=` was faster than `std::stringstream::operator<<` (as I mentioned on the Issue page). My test was rather synthetic but now we have an implementation so @sudo-panda could compare the speed of `bg::wkt` and `bg::to_wkt` and make a case that using `std::string` directly is indeed faster, or not.

My additional reasoning was/is as follows:
1. The simple function above is so trivial that the user could implement it using whatever stream manipulators that he/she needs. If we want to add a function like `to_wkt`, it only makes sense for me if our implementation could in some way beat this trivial user-level implementation, e.g. was faster.
2. If we provide a support for some kind of output then we should use the native way of using this kind of output.
3. Adding a layer of abstraction above appending data to streams and strings is of little complexity WRT what we already have in the library.

Reply:
So which design are we going with?

Reply:
> 
> 
> So which design are we going with?

@awulkiew  thanks for your explanation. I remember now you mentioned this before, but I didn't realize it would have these consequences.

@sudo-panda could you compare the speed between the two implementations as @awulkiew  suggested?

Given all discussions, if you are all in favour of this new implementation (but there were still some discussions about the significant digits), I'm not against and you can go ahead.

Reply:
Sure. I would love to benchmark them.

Reply:
@sudo-panda could you also benchmark alternative implementations of appending coordinates?

If you want your benchmark to be quite full you could consider checking at least the following:
```
std::stringstream ss;
ss.precision(std::numeric_limits<T>::digits10);
ss << val;
out += ss.str();
```
and
```
std::stringstream ss;
// without precision
ss << val;
out += ss.str();
```
and
```
out += boost::lexical_cast<std::string>(val);
```
and
```
out += std::to_string(val);
```

All of the above for:
1. geometry containing only coordinates having little number of significant digits, e.g. 0, 1, etc.
2. geometry containing only coordinates having many significant digits, e.g. 0.333333333333, 0.666666666667, 1/3, 2/3, sqrt(2), pi, etc.

All of the above for geometries having less coordinates/points and more in order to check different coordinate/character ratios in WKT strings, so e.g. to have several MULTIPOINTs containing 1, 10, 100 and 1000 points (at least 2 for 1 and 100?).

So at least 16 combinations.

Reply:
### Result:

I made a heat map on plot.ly. You can access it [here](https://plot.ly/~sudo-panda/3/).

Method Used | No of Points | Time Taken
-------------------|------------------|----------------
stringstream w precision, < sig dgts | 1 points| 0
stringstream w precision, > sig dgts | 1 points| 3.2
stringstream w precision, < sig dgts | 10 points| 26.86
stringstream w precision, > sig dgts | 10 points| 26.16
stringstream w precision, < sig dgts | 100 points| 249.81
stringstream w precision, > sig dgts | 100 points| 265.87
stringstream w precision, < sig dgts | 1000 points| 2496.11
stringstream w precision, > sig dgts | 1000 points| 2616.56
stringstream w precision, < sig dgts | 10000 points| 24895.1
stringstream w precision, > sig dgts | 10000 points| 26237.3
stringstream w/o precision, < sig dgts | 1 points| 0
stringstream w/o precision, > sig dgts | 1 points| 3.37
stringstream w/o precision, < sig dgts | 10 points| 66.75
stringstream w/o precision, > sig dgts | 10 points| 59.92
stringstream w/o precision, < sig dgts | 100 points| 273.41
stringstream w/o precision, > sig dgts | 100 points| 219.3
stringstream w/o precision, < sig dgts | 1000 points| 2625.64
stringstream w/o precision, > sig dgts | 1000 points| 2816.33
stringstream w/o precision, < sig dgts | 10000 points| 21972.1
stringstream w/o precision, > sig dgts | 10000 points| 17633.8
lexical cast, < sig dgts | 1 points| 0
lexical cast, > sig dgts | 1 points| 1.73
lexical cast, < sig dgts | 10 points| 15
lexical cast, > sig dgts | 10 points| 64.22
lexical cast, < sig dgts | 100 points| 600.01
lexical cast, > sig dgts | 100 points| 379.06
lexical cast, < sig dgts | 1000 points| 2525.91
lexical cast, > sig dgts | 1000 points| 2541.71
lexical cast, < sig dgts | 10000 points| 19197.2
lexical cast, > sig dgts | 10000 points| 19841.5
to_string, < sig dgts | 1 points| 0
to_string, > sig dgts | 1 points| 1.02
to_string, < sig dgts | 10 points| 11.17
to_string, > sig dgts | 10 points| 9.03
to_string, < sig dgts | 100 points| 129.17
to_string, > sig dgts | 100 points| 125.14
to_string, < sig dgts | 1000 points| 1065.07
to_string, > sig dgts | 1000 points| 973.36
to_string, < sig dgts | 10000 points| 9447.4
to_string, > sig dgts | 10000 points| 8865.27



> ***N.B.:*** Values for `< sig dgts` with just *one point* was giving time taken values as *9.41413e+11* so i set them to *0*.

The code used to benchmark the methods :
``` C++
#include <iostream>
#include <ctime>
#include <cstdlib>
#include <cmath>
#include <chrono> 
#include <boost/geometry.hpp>
#include <boost/geometry/geometries/point_xy.hpp>
#include <boost/geometry/geometries/multi_point.hpp>

int impr_to_wkt(int no_of_points)
{
    namespace geom = boost::geometry;
    typedef geom::model::d2::point_xy<double> point_type;

    geom::model::multi_point<point_type> multi_pt;
    for(int i = 0; i < no_of_points; i++)
    {
        srand( unsigned(i*time(NULL)));
        geom::append(multi_pt, geom::make<point_type>(rand() % 1000 / 100.0, rand() % 1000 / 100.0));
    }
        

    std::string out;
    std::chrono::_V2::system_clock::time_point start = std::chrono::high_resolution_clock::now();
    out = geom::to_wkt(multi_pt);
    std::chrono::_V2::system_clock::time_point stop = std::chrono::high_resolution_clock::now();
    std::cerr<<out.length()<<"\t";

    return std::chrono::duration_cast<std::chrono::microseconds>(stop - start).count();
}

int prec_to_wkt(int no_of_points)
{
    namespace geom = boost::geometry;
    typedef geom::model::d2::point_xy<double> point_type;

    geom::model::multi_point<point_type> multi_pt;
    for(int i = 0; i < no_of_points; i++)
    {
        srand( unsigned(i*time(NULL)));
        geom::append(multi_pt, geom::make<point_type>((double)rand()/(RAND_MAX), (double)rand()/(RAND_MAX)));
    }
    
    std::string out;
    std::chrono::_V2::system_clock::time_point start = std::chrono::high_resolution_clock::now();
    out = geom::to_wkt(multi_pt);
    std::chrono::_V2::system_clock::time_point stop = std::chrono::high_resolution_clock::now();
    std::cerr<<out.length()<<"\t";

    return std::chrono::duration_cast<std::chrono::microseconds>(stop - start).count();
}

int main()
{
    for(int i = 0; i < 5; i++)
    {
        long long sum;
        for(int j = 0 ; j<100; j++)
        {
            sum += impr_to_wkt(pow(10, i));
        }
        std::cout<<"stringstream w precision, "<<" < sig dgts, "<<(int) pow(10, i)<<" points, "<<sum/100.0<<std::endl;
        sum = 0;
        for(int j = 0 ; j<100; j++)
        {
            sum += prec_to_wkt(pow(10, i));
        }
        std::cout<<"stringstream w precision, "<<" > sig dgts, "<<(int) pow(10, i)<<" points, "<<sum/100.0<<std::endl;
    }
}
```

### Inference:
Time taken by `to_string` is the least followed by `lexical_cast`. The `stringstream` method is the slowest When the precision is set, it makes it slower.

Reply:
@sudo-panda thanks! I don't see the benchmark for `bg::wkt()`. Have you tested it too?

Do I see correctly that stringstream with precision set to digits10 is faster than the one without it? That's wierd.

Reply:
What compiler version and options have you used?

Reply:
> I don't see the benchmark for `bg::wkt()`. Have you tested it too?

I haven't tested it.

> Do I see correctly that stringstream with precision set to digits10 is faster than the one without it? That's wierd.

No `stringstream` with precision is slower than the one without as I said:

>Time taken by `to_string` is the least followed by `lexical_cast`. The `stringstream` method is the slowest When the precision is set, it makes it slower.

It can also be seen in the heat map provided.

> What compiler version and options have you used?

I have used the default `b2` configuration which uses `gcc-7` in my case. As for the options can you tell me how can I find that?

Reply:
@sudo-panda
>  As for the options can you tell me how can I find that?

You definitally want to benchmark **optimised** build, not debug (which is default), that is:

```
b2 variant=release address-model=32
b2 variant=release address-model=64
```

More options at https://boostorg.github.io/build/manual/develop/index.html#bbv2.overview.invocation.properties

Reply:
@sudo-panda

> No stringstream with precision is slower

Only for geometries with 10000 points and 100 > sig dgts.

> I have used the default b2 configuration

Ok, so the results are irrelevant for now.

I suggest to NOT use b2 at all and to build your program directly with a compiler and pass -O2 or -O3 option explicitly.

Reply:
> I suggest to NOT use b2 at all and to build your program directly with a compiler and pass -O2 or -O3 option explicitly.

I'm not advocating for b2, but for completeness of the picture, it is also allows that:

```
b2 variant=release address-model=64 optimization=space
b2 variant=release address-model=64 optimization=speed
b2 variant=release address-model=64 optimization=full
```

If you want to see actual compiler command line invoked, just add `-d2 as in `b2 -d2 variant=release ....`

Reply:
I have added `wkt` and `to_wkt` operations to my string benchmark in https://github.com/mloskot/string_benchmark/blob/master/benchmark_boost_geometry_wkt.cpp, where

- `std::to_string` is the baseline
- problem space means number of points (pairs of X/Y)

### GCC 9

CSV files with results: https://github.com/mloskot/string_benchmark/blob/master/results/gcc9_geometry_wkt.csv

![image](https://user-images.githubusercontent.com/80741/77376552-decd9080-6d70-11ea-8987-209005401656.png)

### MSVC++ 14.25 (VS2019)

CSV files with results: https://github.com/mloskot/string_benchmark/blob/master/results/vs2019_geometry_wkt.csv

![image](https://user-images.githubusercontent.com/80741/77376810-8f3b9480-6d71-11ea-9be6-b64d88dc69fb.png)

------

I picked up @awulkiew's theme on speeding things up because I think speed is important here. Adam's suggestion to re-work the internals opened possibility to completely avoid using streams for `to_wkt` and use of `std:;string` only, even with size `reserve`-ed heuristically based on number of points and dimension of geometry to further tweak performance (along with measuring it, of course). I hope this helps.

Reply:
What is the difference in `wkt` and `stringstream` experiments?

Reply:
@sudo-panda
> What is the difference in `wkt` and `stringstream` experiments?

Check the very _trivial_ code of [benchmark_boost_geometry_wkt.cpp](https://github.com/mloskot/string_benchmark/blob/master/benchmark_boost_geometry_wkt.cpp), you will see immediately. The `to_string`, `stringstream` and `lexical_cast` are generic cases outputting **N** points then `wkt` and `to_wkt` do similar for `multi_point` of **N** points, where  **N** is the **Problem Space** number of points (pairs of X/Y). When you compare results, compare test cases of the the same problem spaces.

Reply:
Ok I got confused thinking that you were using them in the `output_formatter` rather than benchmarking then standalone.

Reply:
No, I did not apply any modifications to the Boost.Geometry (code of this PR) at all.

I just wrote quick benchmark which just compares

1. 3 x basic general-purpose methods to output list of `double`-s to string methods: `to_string`, `stringstream` and `lexical_cast`
2. 2 x Boost.Geometry methods: `wkt` and `to_wkt`

I think of the 1. as reference results and I picked `to_string` as the single baseline to contrast all other methods against..
I think of the 2. as results of interest for further optimisation.

So, next, the benchmark could be completed with the Geometry's methods further tweaked, optimised, etc.

Reply:
Sorry for the late reply.

I tried celero for the first time and the tool is really great although I don't know what to make of the varied results I got.

I have uploaded my code to my repo: https://github.com/sudo-panda/bg_wkt_benchmark

The compiler I have used is GNU C++ 7.5.0.

I have also tried both O2 and O3 and an unoptimized version(as stated in celero docs)

You can find the results of the respective operations in the results directory of my repo. I have also formatted it and sorted it in a google sheet and you can find that here: https://github.com/sudo-panda/bg_wkt_benchmarkhttps://docs.google.com/spreadsheets/d/1AYqr0Unob5Ee6_-6ThEg3Z3iROzOuE01tmLQ46_JJis/edit?usp=sharing

Reply:
@sudo-pand I didjust quick look, but I'm not sure.
I would suggest you to let Celero do the statistics, remove `30` and stick `0`, then bump the `1` to more meaningful size.  You will also notice that I deliberately use `0` for number of iterations in `ExperimentValues`. Another thing is that I'm not sure about Celero's statistics when you force baseline to no-op. If you want to control benchmark with hard-wired runs, iterations, etc. I'd suggest you to use https://github.com/google/benchmark or https://github.com/nickbruun/hayai instead. Benchmarking needs time. My variant is simple and I think it gives quite clear results, but it does not include all test cases we may be interested in.

Reply:
I made the changes you suggested but I still can't find the method that is clearly better than the others. The results are a bit mixed. The changes are uploaded to the develop branch. 

https://github.com/sudo-panda/bg_wkt_benchmark/tree/develop

I also updated the google sheet with new results:

https://docs.google.com/spreadsheets/d/1AYqr0Unob5Ee6_-6ThEg3Z3iROzOuE01tmLQ46_JJis/edit?usp=sharing

Reply:
@sudo-panda I posted benchmarks update in the main thread of comments here https://github.com/boostorg/geometry/pull/670#issuecomment-620277337